main/stm32bl.c

```diff
@@ -86,11 +114,46 @@ int stm32bl_go(uint32_t addr)
     if (wait_resp() != STM32BL_ACK) {
         return ESP_FAIL;
     }
-
     send_address(addr);
     return wait_resp();
 }
 
+static int stm32bl_run_raw_code(uint32_t target_address, const uint8_t *code, uint32_t code_size)
+{
+	uint32_t stack_le = 0x20002000;
+	uint32_t code_address_le = target_address + 8 + 1; // thumb mode address (!)
+	uint32_t length = code_size + 8;
+	uint8_t *mem, *pos;
+	uint32_t address;
+
+	/* Must be 32-bit aligned */
+	if (target_address & 0x3) {
+		ESP_LOGE(TAG, "Error: code address must be 4 byte aligned\n");
+		return ESP_FAIL;
+	}
+
+	mem = malloc(length);
+	if (!mem)
+		return ESP_FAIL;
+
+	memcpy(mem, &stack_le, sizeof(uint32_t));
+	memcpy(mem + 4, &code_address_le, sizeof(uint32_t));
+	memcpy(mem + 8, code, code_size);
+
+	pos = mem;
+	address = target_address;
+    stm32bl_write(pos, length, address);
+
+	free(mem);
+	return stm32bl_go(target_address);
+}
+
+int stm32bl_reset_device()
+{
+    uint32_t address = 0x20002000;
```

Comment:
Also here, if this is a contant, I suggest to define it in the header and name it properly so that we don't have a magic number here in the middle of the code.